rfcs/2021-03-26-6517-end-to-end-acknowledgement.md

```diff
@@ -0,0 +1,344 @@
+leave this behaviour in the hands of script writers, with Vector
+providing such scripts with convenience library functions to assist with
+the common tasks outlined above.
+
+### Non-acknowledging source
+
+Some sources are unable to provide acknowledgements at a protocol level
+(for example, `apache_metrics`, `prometheus_parser`, or `stdin`).
+Further, users of Vector may wish to selectively enable this feature
+only selectively on certain sources. To minimize the overhead of this
+feature, sources that cannot or should not provide acknowledgements
+should not be required to participate in finalization handling, as they
+would only be discarding the status.
+
+### Sinks using a disk buffer
+
+In order to deal with longer delivery delays and backpressure, sinks may
+be configured to store events temporarily in a disk buffer. Since the
+events are not actually delivered when they are buffered, the final
+delivery status will have to be confirmed after the event is loaded back
+from the buffer. The source, to be used to determine where
+to provide the finalization status indicator, must be converted into a form
+that may be serialized (ie string or integer index). However, the
+configuration may be reloaded while the event is buffered in such a way
+that the source name provided in the configuration may change, or a
+different component may be substituted for the same identifier. As such,
+the configured component name is not sufficient to uniquely identify the
+source. Further, Vector may be stopped while the event is buffered and
+then restarted with an identical configuration. At this point, the old
+source will be distinct from the current source, despite having the same
+configured name. As such, the persisted identifier for the source must
+stay the same across configuration reloads, but must be different for
+each run of Vector.
+
+## Internal Proposal
+
+There are two components to the proposal: the communication between
+event finalization and the sources, and the resulting event
+metadata required to track the event status.
+
+### Event finalization
+
+When instantiated, the topology will provide each source with a unique
+identifier for the token. The source may use this identifier to create
+one or more notification channels, each with their own unique
+identifier. This provides both a unique identifier of the source that
+needs to receive the event finalization status as well as the mechanism
+for delivering the event identifier and status.
+
+The finalization status is recorded when the last copy of the event is
+dropped. At that time, the finalization status stored in the event is
+recorded in a batch status buffer. Since events may be internally
+dropped when they are merged, a special "no-op" status is used to
+indicate that no status needs to be recorded to prevent superfluous
+status updates. When the last copy of the batch status is dropped, it is
+sent to the source.
```

Comment:
I have a question about undelivarable events. So, we have some sink that will never ack some percentage of `Event` instances but is otherwise functional, say it's some hostile sink that is written to not ack every fifth `Event`. Presumably these undeliverable `Event`s count toward the back-pressure between the final hop and the sink but does it count for the back-pressure of the source? If so and if the pipeline is not linear can we avoid a hostile sink denying advancement to all other sinks? 

Actually, maybe another way to ask this is in the following situation:

```
source ------------> fast_sink
               |
               ----------------> slow_sink 
```

Does the `slow_sink` cause back-pressure that affects the `fast_sink`?